pkg/virt-controller/watch/vmi.go

```diff
@@ -254,12 +261,46 @@ func (c *VMIController) execute(key string) error {
 
 	// If needsSync is true (expectations fulfilled) we can make save assumptions if virt-handler or virt-controller owns the pod
 	needsSync := c.podExpectations.SatisfiedExpectations(key)
+	volumeStatusCopy := vmi.Status.DeepCopy().VolumeStatus
+	if pod != nil {
+
+		hotplugVolumes := c.getHotplugVolumes(vmi, pod)
+		hotplugAttachmentPods, err := c.virtlauncherAttachmentPods(pod)
+		if err != nil {
+			return err
+		}
+		if pod.DeletionTimestamp == nil && c.needsHandleHotplug(hotplugVolumes, hotplugAttachmentPods) {
+			var hotplugSyncErr syncError = nil
+			volumeStatusCopy, hotplugSyncErr = c.handleHotplugVolumes(hotplugVolumes, hotplugAttachmentPods, vmi, pod)
+			if hotplugSyncErr != nil {
+				if hotplugSyncErr.Reason() == MissingAttachmentPodReason {
+					// We are missing an essential hotplug pod. Delete all pods associated with the VMI.
+					needsSync = true
+					c.deleteAllMatchingPods(vmi)
+				} else {
+					return hotplugSyncErr
+				}
+			}
+		} else {
+			if len(volumeStatusCopy) == 0 {
+				// Populate with permanent volumes
+				for _, volume := range vmi.Spec.Volumes {
+					volumeStatusCopy = append(volumeStatusCopy, virtv1.VolumeStatus{
+						Name: volume.Name,
+					})
+				}
+			}
+		}
+		// Remove any volume statuses that are hotplugged but no longer have an attachment pod.
+		volumeStatusCopy = c.cleanupVolumeStatus(volumeStatusCopy, hotplugAttachmentPods)
+	}
+	logger.V(1).Infof("volumeStatusCopy [%v]", volumeStatusCopy)
```

Comment:
this needs to be split between the sync() function and the updateStatus function. 

Sync is protected by this `needsSync` boolean which ensures all expectations are met before any actions can be performed (like creating/deleting pods)

UpdateStatus observes the actions performed by sync() through the usage of whatever that function observes in the informer cache.